DESIGN_of_proposal.md

```diff
@@ -0,0 +1,354 @@
+# Design Document: The `of` Type System for JuliaBUGS
+
+## Overview
+
+This document outlines the design and implementation of the `of` type system, which replaces the current `@parameters` macro with a more flexible and Julia-idiomatic approach to parameter specification in JuliaBUGS models. The system provides pytree-like composability for complex hierarchical parameter structures commonly found in probabilistic programming.
+
+## Motivation
+
+The current `@parameters` macro has several limitations:
+1. It requires a separate struct definition outside the model
+2. Type annotations are not supported
+3. It doesn't integrate well with Julia's type system
+4. Parameter initialization is awkward with `ParameterPlaceholder`
+5. At least as of Julia v1.11, struct can't be redefined, which makes it difficult to use
+6. No support for hierarchical or nested parameter structures
+
+The `of` system addresses these issues by:
+- Providing inline type specifications
+- Supporting both tuple and named tuple destructuring
+- Offering direct integration with Julia's type system
+- Enabling natural parameter initialization with `rand()` and `zero()`
+- Supporting arbitrary nesting and composition of parameter structures
+- Providing pytree-like utilities for parameter manipulation
+
+## Core Design
+
+### 1. The `of` Type Constructor
+
+The `of` function creates type specifications for BUGS parameters:
+
+```julia
+# Basic usage
+of(Array, dims...)           # Array with specified dimensions
+of(Array, T, dims...)        # Array with element type T
+of(Real)                     # Scalar real number
+of(Real, lower, upper)       # Bounded real number
+```
+
+### 2. Type Representations
+
+The `of` function returns type objects that can be:
+- Used to generate random values: `rand(of(Array, 3, 4))`
+- Used to generate zero values: `zero(of(Real))`
+- Used as type annotations in function signatures
+- Combined into tuples and named tuples
+
+### 3. Parameter Specifications
+
+#### Use Case 1: Runtime Parameter Generation
+```julia
+x = of(Array, 8)
+y = of(Array, of(Real), 4, 3)
+w = of(Real, lower, upper)
+
+# Create parameter tuple/named tuple
+params = of((x, y, w))                    # Tuple
+params = of((x=x, y=y, w=w))             # NamedTuple
+
+# Use in model
+model(rand(params), constants...)
+```
+
+#### Use Case 2: Type Annotations
+```julia
+# Tuple parameters
+Tparameters = Tuple{of(Array, 8), of(Array, of(Real), 4, 3), of(Real, lower, upper)}
+
+@model function demo((x, y, w)::Tparameters, constants...)
+    # model body
+end
+
+# NamedTuple parameters
+NTparameters = @NamedTuple{x::of(Array, 8), y::of(Array, of(Real), 4, 3), w::of(Real, lower, upper)}
+
+@model function demo((;x, y, w)::NTparameters, constants...)
+    # model body
+end
+```
+
+#### Use Case 3: Direct Support in `@model`
+```julia
+# Tuple syntax
+@model function demo(
+    (x::of(Array, 8), y::of(Array, of(Real), 4, 3), w::of(Real, lower, upper)),
```

Comment:
One issue with such tuple syntax is variables `x`, `y`, etc won't be available in the function body since Julia sees the first argument `(x, y, ...)` as a single argument by default. We can override this behaviour, but that might surprise users.